### PR TITLE
move unbind to ATen

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -608,5 +608,15 @@ int64_t numel(const Tensor& self) {
   return self.pImpl->numel();
 }
 
+std::vector<Tensor> unbind(const Tensor &self, int64_t dim) {
+  AT_CHECK(dim < self.dim(), "dim must be smaller than ", self.dim());
```
Comment (apaszke) [inline aten/src/ATen/native/TensorShape.cpp]: Please use `maybe_wrap_dim`

**Review by apaszke — COMMENTED** [review]
Can you please also add the function to `tools/autograd/derivatives.yaml` and specify that its derivative is `stack`? This will be much more efficient than what's currently generated by this op. Thanks!

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -5719,3 +5719,17 @@ def parse_kwargs(desc):
...
+unbind(tensor, dim=0) -> seq
+
+Removes a tensor dimension.
+
+Returns a tuple of all slices along a given dimension, already without it.
+
+Arguments:
+    tensor (Tensor): the tensor to unbind
+    dim (int): dimension to remove
+""")
```
Comment (vishwakftw) [inline torch/_torch_docs.py]: Can you provide an example too?

**Comment by zasdfgbnm** [comment]
@apaszke @vishwakftw Done

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -48,6 +48,10 @@ void copy_range(variable_list& out, IndexRange range, at::ArrayRef<Tensor> t) {
   std::copy(t.begin(), t.end(), out.begin() + range.first);
 }
 
+std::vector<Tensor> to_tensor_list(const variable_list& variables) {
+  return fmap(variables, [](torch::autograd::Variable v) { return static_cast<Tensor>(v); } );
```
Comment (apaszke) [inline tools/autograd/templates/Functions.cpp]: Please make it `const torch::autograd::Variable& v`. Right now it incurs an extra temporary refcount increase, which is unnecessary

File: aten/src/ATen/native/TensorShape.cpp [hunk 4]
```diff
@@ -608,5 +608,15 @@ int64_t numel(const Tensor& self) {
...
 
+std::vector<Tensor> unbind(const Tensor &self, int64_t dim) {
+  dim = maybe_wrap_dim(dim, self.dim());
+  int64_t size = self.size(dim);
+  std::vector<Tensor> tensors(size);
+  for (int i = 0; i < size; i++) {
+    tensors[i] = self.select(dim, i);
+  }
+  return tensors;
+}
```
Comment (apaszke) [inline aten/src/ATen/native/TensorShape.cpp]: @colesbury do you remember if we need to do anything special to make outputs to be treated as views? I'm not sure how the fact that they come from `select` internally plays with the custom gradient definition

**Comment by zou3519** [comment]
cc @apaszke how does this look now?

File: aten/src/ATen/native/TensorShape.cpp (hunk 4 above)
Comment (colesbury) [reply aten/src/ATen/native/TensorShape.cpp]: Yes, it needs to be added to VIEW_FUNCTIONS in `gen_variable_type.py`, but that probably does not support std::vector<Tensor>

Comment (zasdfgbnm) [reply aten/src/ATen/native/TensorShape.cpp]: @apaszke @colesbury I have added it to `VIEW_FUNCTIONS` and add an overload of `as_view` for vector of tensors.

**Review by apaszke — COMMENTED** [review]
LGTM! Can you please add grad tests for unbind (it has its own grad definition now)?

**Comment by apaszke** [comment]
Another thing I'm wondering about is if we won't end up paying the cost of creating autograd functions for those nested `select`s even though the grad will later be overriden by the outer `UnbindBackward`. Shouldn't we enable no-grad mode in `VariableType.cpp` before we call into the native function? @colesbury?

**Comment by zasdfgbnm** [comment]
@apaszke Isn't the grad test already there?
https://github.com/pytorch/pytorch/pull/8587/files#diff-e7525582ab432c30821289d64cc52bc3R1287

**Review by apaszke — APPROVED** [review]

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**